eigenpro/solver.py

```diff
@@ -132,21 +129,18 @@ def fit(model, X, Y, x, y, device, dtype=torch.float32, kernel=None,
                 torch.cuda.empty_cache()
 
             if ( (project_counter + 1) % T == 0 or (t==len(train_dataloader)-1) ) and accumulated_gradients:
-                projection_dataset = array_dataset.ArrayDataset(model.centers[0], optimizer.grad_accumulation)
-                projection_loader = torch_data.DataLoader(
-                    projection_dataset, batch_size=model_batch_size, shuffle=True)
-                for _ in range(1):
-                    for z_batch, grad_batch, id_batch in tqdm(
-                            projection_loader, 
-                            total=len(projection_loader), 
-                            leave=False,
-                            desc="Projection"
-                        ):
-                        optimizer.step(z_batch, grad_batch, id_batch, projection=True)
-
-                update_projection = torch.cat([k.weights_project.to(device_base) for k in model.shard_kms])
-
-                model.update_by_index(torch.tensor(list(range(p))), update_projection)
+
+                if model_preconditioner is None:
```

Comment:
I am now inclined to define a `Projector` class to
hide `nys_model_indices` and the following preconditioner, which is really part of the projection

what do you think?

Reply:
I think since the nystrom_indices also gonna change every time that we update the preconditioner, there is a little point to make it a class

Reply:
This is only for hiding the implementation complexity. Unfortunately, there is not going to have any computation saving.